virttest/utils_disk.py

```diff
@@ -600,6 +600,103 @@ def create_filesystem(session, partition_name, fstype, ostype, timeout=360):
         create_filesyetem_linux(session, partition_name, fstype, timeout)
 
 
+def get_fstype_mpoint_linux(session, partition):
+    """
+    Get a partition's file system type and mountpoint.
+
+    :param session: session object to guest.
+    :param partition: disk partition, like /dev/sdb1.
+    :return: Tuple (fstype, mountpoint)
+    """
+    mount_list = session.cmd_output('cat /proc/mounts')
+    pattern = r'%s\s(.+?)\s(.+?)\s' % partition
+    mount_info = re.search(pattern, mount_list)
+    if mount_info:
+        mountpoint = mount_info.group(1)
+        fstype = mount_info.group(2)
+        return (fstype, mountpoint)
+    logging.error('partition should not be mounted, mounted'
+                  'list as:\n%s' % mount_list)
+
+
+def resize_filesystem_linux(session, partition, size):
+    """
+    resize file system in linux guest.
+    support enlarge and shrink.
+
+    :param session: session object to guest.
+    :param partition: disk partition, like /dev/sdb1.
+    :param size: resize file system to size, unit is B.
+                 support transfer size='MAX_AVAILABLE_SIZE',
+                 enlarge to maximun available size.
+    """
+    fstype, mountpoint = get_fstype_mpoint_linux(session, partition)
+    if fstype == 'xfs':
+        if size == 'MAX_AVAILABLE_SIZE':
+            resize_fs_cmd = 'xfs_growfs -d %s' % mountpoint
+        else:
+            get_bsize_cmd = 'xfs_growfs -n %s' % mountpoint
+            output = session.cmd_output(get_bsize_cmd)
+            bsize = re.findall(r'data\s+=\s+bsize=(\d+)', output, re.M)[0]
+            blocks = int(size.split(".")[0]) / int(bsize) - 512
+            resize_fs_cmd = 'xfs_growfs -D %s %s' % (blocks, mountpoint)
+    else:
+        if is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+            umount(partition, mountpoint, fstype=fstype, session=session)
+
+        session.cmd('e2fsck -f %s' % partition)
+
+        if size == 'MAX_AVAILABLE_SIZE':
+            resize_fs_cmd = 'resize2fs %s' % partition
+        else:
+            size = utils_numeric.normalize_data_size(size, order_magnitude="M")
+            size = int(size.split(".")[0]) - 2
+            resize_fs_cmd = 'resize2fs %s %sM' % (partition, size)
+    session.cmd(resize_fs_cmd)
+
+    if not is_mount(partition, dst=mountpoint, fstype=fstype, session=session):
+        mount(partition, mountpoint, fstype=fstype, session=session)
+
+
+def resize_partition_linux(session, partition, size, start='0M'):
```

Comment:
Some errors:
```
>>> print(session.cmd('lsblk'))
NAME                      MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sda                         8:0    0   20G  0 disk 
├─sda1                      8:1    0    4M  0 part 
├─sda2                      8:2    0    1G  0 part /boot
└─sda3                      8:3    0   19G  0 part 
  ├─rhel_dhcp71--240-root 253:0    0   17G  0 lvm  /
  └─rhel_dhcp71--240-swap 253:1    0    2G  0 lvm  [SWAP]
sdb                         8:16   0  100G  0 disk 
└─sdb1                      8:17   0 94.4G  0 part /mnt/sdb1

>>> utils_disk.resize_partition_linux(session, '/dev/sdb1', '5000M', start='0M')
Traceback (most recent call last):
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1130, in cmd_output
    o = self.read_up_to_prompt(timeout, internal_timeout, print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1099, in read_up_to_prompt
    print_func)[1]
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 925, in read_until_last_line_matches
    print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 843, in read_until_output_matches
    raise ExpectTimeoutError(patterns, o)
aexpect.exceptions.ExpectTimeoutError: Timeout expired while looking for pattern '^\\[.*\\][\\#\\$]\\s*$'    (output: 'e2fsck 1.44.3 (10-July-2018)\nError reading block 12091392 (Invalid argument).  Ignore error<y>? ')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/yhong/kar/workspace/avocado-vt/virttest/utils_disk.py", line 695, in resize_partition_linux
    session.cmd('e2fsck -f %s' % partition)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1286, in cmd
    print_func)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1213, in cmd_status_output
    o = self.cmd_output(cmd, timeout, internal_timeout, print_func, safe)
  File "/home/yhong/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 1134, in cmd_output
    raise ShellTimeoutError(cmd, o)
aexpect.exceptions.ShellTimeoutError: Timeout expired while waiting for shell command to complete: 'e2fsck -f /dev/sdb1'    (output: 'e2fsck 1.44.3 (10-July-2018)\nError reading block 12091392 (Invalid argument).  Ignore error<y>? ')
>>> 

```

Reply:
As we discussed, this function is connection with  resize_filesystem_linux(), in order to reduce user misunderstand this connection, so I packaged enlarge_disk_size_linux() and shrink_disk_size_linux() functions as well.